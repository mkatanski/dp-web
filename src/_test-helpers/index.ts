export * from "./withResponsiveContext";
export * from "./withThemeProvider";
