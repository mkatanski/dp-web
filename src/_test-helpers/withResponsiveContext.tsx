import React from "react";
import { Context as ResponsiveContext } from "react-responsive";
import { Breakpoints } from "config/breakpoints";

export const withMobileSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.mobile }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withTabletSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.tablet }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.desktop }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withLargeDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.largeDesktop }}>
    {Component}
  </ResponsiveContext.Provider>
);
