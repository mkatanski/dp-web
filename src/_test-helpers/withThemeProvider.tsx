import React from "react";
import { ThemeProvider as DefaultThemeProvider } from "lib/theme/ThemeProvider";
import { StyledTheme, Theme } from "styles/theme/_types";
import { createMuiTheme } from "@material-ui/core";

const testTheme: StyledTheme = {
  primary: "primary",
  secondary: "secondary",
  textPrimary: "textPrimary",
  textSecondary: "textSecondary",
  textInverse: "textInverse"
};

const muiTheme = createMuiTheme({
  direction: "ltr"
});

const theme: Theme = Object.assign(testTheme, muiTheme);

export const withThemeProvider = (Component: React.ReactNode) => (
  <DefaultThemeProvider theme={theme}>{Component}</DefaultThemeProvider>
);
