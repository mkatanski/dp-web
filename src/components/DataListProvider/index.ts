export * from "./DataListProvider";
export * from "./DataListContext";
