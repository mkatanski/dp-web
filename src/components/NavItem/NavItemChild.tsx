import React from "react";
import { NavLink as RouterLink } from "react-router-dom";
import { Button, ListItem, SvgIconProps } from "@material-ui/core";
import styled from "styled-components";

const DEFAULT_SPACING = 8;
const LINK_INDENT = 32;

export type NavItemChildProps = {
  title: string;
  href: string;
  depth: number;
  icon?: React.ReactType<SvgIconProps>;
  info?: React.ReactNode;
};

const ListItemWrapper = styled.span`
  .itemLeaf {
    display: flex;
    padding-top: 0;
    padding-bottom: 0;
  }

  .linkButtonLeaf {
    color: ${({ theme }) => theme.palette.text.secondary};
    padding: ${({ theme }) => theme.spacing(1.25, 1)};
    justify-content: flex-start;
    text-transform: none;
    letter-spacing: 0;
    width: 100%;
    font-weight: ${({ theme }) => theme.typography.fontWeightRegular};

    &.depth-0 {
      &.title {
        font-weight: ${({ theme }) => theme.typography.fontWeightMedium};
      }
    }
  }

  .linkTitle {
    margin-right: auto;
  }

  .icon {
    display: flex;
    align-items: center;
    margin-right: ${({ theme }) => `${theme.spacing(1)}px`};
  }

  .linkActive {
    color: ${({ theme }) => theme.palette.secondary.main};

    .linkTitle {
      font-weight: ${({ theme }) => theme.typography.fontWeightMedium};
    }

    .icon {
      color: ${({ theme }) => theme.palette.secondary.main};
    }
  }
`;

export const NavItemChild: React.FC<NavItemChildProps> = ({
  title,
  href,
  depth,
  icon: Icon,
  info: Info
}: NavItemChildProps) => {
  let paddingLeft = DEFAULT_SPACING;

  depth > 0 && (paddingLeft = LINK_INDENT + DEFAULT_SPACING * depth);

  const style = { paddingLeft };

  return (
    <ListItemWrapper>
      <ListItem className="itemLeaf" disableGutters key={title}>
        <Button
          component={RouterLink}
          activeClassName="linkActive"
          className={`linkButtonLeaf depth-${depth}`}
          style={style}
          to={href}
        >
          {Icon && <Icon className="icon" fontSize="default" />}
          <span className="linkTitle">{title}</span>
          {Info}
        </Button>
      </ListItem>
    </ListItemWrapper>
  );
};
