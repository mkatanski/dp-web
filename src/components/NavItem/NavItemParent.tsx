import React, { useState } from "react";
import { Button, Collapse, ListItem, SvgIconProps } from "@material-ui/core";
import { ExpandMoreOutlined, ExpandLessOutlined } from "@material-ui/icons";
import styled from "styled-components";

export type NavItemParentProps = React.PropsWithChildren<{
  title: string;
  isOpen?: boolean;
  icon?: React.ReactType<SvgIconProps>;
}>;

const ListItemWrapper = styled.span`
  .item {
    display: block;
    padding-top: 0;
    padding-bottom: 0;
  }

  .linkButton {
    color: ${({ theme }) => theme.palette.text.secondary};
    padding: ${({ theme }) => theme.spacing(1.25, 1)};
    justify-content: flex-start;
    text-transform: none;
    letter-spacing: 0;
    width: 100%;
  }

  .linkTitle {
    margin-right: auto;
  }

  .icon {
    display: flex;
    align-items: center;
    margin-right: ${({ theme }) => `${theme.spacing(1)}px`};
  }
`;

export const NavItemParent: React.FC<NavItemParentProps> = ({
  title,
  icon: Icon,
  isOpen: isOpenProp = false,
  children
}: NavItemParentProps) => {
  const [isOpen, setOpen] = useState(isOpenProp);

  const handleToggle = () => {
    setOpen(prevOpen => !prevOpen);
  };

  return (
    <ListItemWrapper>
      <ListItem className="item" disableGutters key={title}>
        <Button className="linkButton" onClick={handleToggle}>
          {Icon && <Icon className="icon" fontSize="default" />}
          <span className="linkTitle">{title}</span>
          {isOpen ? (
            <ExpandLessOutlined fontSize="default" color="inherit" />
          ) : (
            <ExpandMoreOutlined fontSize="default" color="inherit" />
          )}
        </Button>
        <Collapse in={isOpen}>{children}</Collapse>
      </ListItem>
    </ListItemWrapper>
  );
};
