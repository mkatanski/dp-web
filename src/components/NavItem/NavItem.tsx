import React from "react";

import { NavItemParent, NavItemParentProps } from "./NavItemParent";
import { NavItemChild, NavItemChildProps } from "./NavItemChild";

export type NavItemProps = NavItemParentProps & NavItemChildProps;

export const NavItem: React.FC<NavItemProps> = ({
  children,
  ...rest
}: NavItemProps) =>
  children ? (
    <NavItemParent {...rest}>{children}</NavItemParent>
  ) : (
    <NavItemChild {...rest} />
  );
