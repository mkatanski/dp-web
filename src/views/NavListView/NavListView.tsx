import React from "react";
import { paths } from "config/paths";
import { List } from "@material-ui/core";
import { Dashboard } from "@material-ui/icons";
import { NavItem } from "components/NavItem";
import styled from "styled-components";

const NavListWrapper = styled.div`
  height: 100%;
  display: flex;
  flex-direction: column;
`;
const Nav = styled.nav`
  overflow: auto;
  padding: ${({ theme }) => theme.spacing(0, 2, 2, 2)};
  flex-grow: 1;
`;
export const NavListView: React.FC = () => (
  <NavListWrapper>
    <Nav>
      <NavItem title="General" depth={0} icon={Dashboard} href="" isOpen>
        <List>
          <NavItem title="Deployments" href={paths.deployments} depth={1} />
          <NavItem title="Templates" href={paths.templates} depth={1} />
        </List>
      </NavItem>
    </Nav>
  </NavListWrapper>
);
