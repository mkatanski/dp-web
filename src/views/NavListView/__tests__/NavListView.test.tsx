/* eslint-disable react/display-name */
import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";

import { NavListView } from "../NavListView";
import { NavItemProps } from "components/NavItem";
import { withThemeProvider } from "_test-helpers";

jest.mock("components/NavItem", () => ({
  NavItem: ({ ...rest }: React.FC<NavItemProps>) => (
    <div id="NavItem" data-prop={rest} />
  )
}));

describe("views", () => {
  describe("NavListView", () => {
    it("should render NavListView component", () => {
      const tree = TestRenderer.create(withThemeProvider(<NavListView />));

      expect(tree).toMatchSnapshot();
    });
  });
});
