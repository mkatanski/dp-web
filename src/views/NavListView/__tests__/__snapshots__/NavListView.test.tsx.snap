// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views NavListView should render NavListView component 1`] = `
.c0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  overflow: auto;
  padding: 0px 16px 16px 16px;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

<div
  className="c0"
>
  <nav
    className="c1"
  >
    <div
      data-prop={
        Object {
          "children": <List>
            <NavItem
              depth={1}
              href="/deployments"
              title="Deployments"
            />
            <NavItem
              depth={1}
              href="/templates"
              title="Templates"
            />
          </List>,
          "depth": 0,
          "href": "",
          "icon": [Function],
          "isOpen": true,
          "title": "General",
        }
      }
      id="NavItem"
    />
  </nav>
</div>
`;
